beard/clustering/blocking_funcs.py

```diff
@@ -0,0 +1,487 @@
+    on some of the signatures, the new signature will be assigned to the block
+    of the last surname.
+
+    Otherwise, the algorithm check how many signatures have the same first
+    names or initials for both considered blocks. The numbers are normalized
+    using sizes of the blocks and compared with each other. The new signature
+    is assigned to the block with bigger score.
+
+    To prevent creation of too big clusters, the ``threshold`` parameter can
+    be set. The algorithm will split every block which size is bigger than
+    ``threshold`` into smaller ones using first names initials as the
+    condition.
+
+    Parameters
+    ----------
+    :param X: numpy array
+        Array of one element arrays of dictionaries. Each dictionary
+        represents a signature. The algorithm needs ``author_name`` field in
+        the dictionaries in order to work.
+    :param threshold: integer
+        Size above which the blocks will be split into smaller ones.
+
+    Returns
+    -------
+    :returns: numpy array
+        Array with ids of the blocks. The ids are strings. The order of the
+        array is the same as in the ``X`` input parameter.
+    """
+    # Stores all clusters. It is the only way to access them.
+    # Every cluster can be accessed by the token that was used to create it.
+    # It is the last token from the surnames tokens passed to the constructor.
+    id_to_cluster = {}
+
+    # List of tuples. Used as the in-between state of the algorithm between
+    # the first and the second states. The tuple contain the block name
+    # if the signature has been already blocked or None otherwise, and the
+    # tokens.
+    ordered_tokens = []
+
+    # First phase.
+    # Create blocks for signatures with single surname
+
+    for signature_array in X[:, 0]:
+        tokens = dm_tokenize_name(signature_array['author_name'])
+        surname_tokens = tokens[0]
+        if len(surname_tokens) == 1:
+            # Single surname case
+            last_name = surname_tokens[0]
+            if last_name not in id_to_cluster:
+                id_to_cluster[last_name] = _Block(*tokens)
+            else:
+                id_to_cluster[last_name].add_signature(*tokens)
+            ordered_tokens.append((last_name, tokens))
+        else:
+            # Multiple surnames
+            ordered_tokens.append((None, tokens))
+
+    # Second phase.
+    # Assign every signature with multiple surnames to the block of the
+    # first surname or the block of the last surname.
+
+    blocks = []
+
+    for token_tuple in ordered_tokens:
+
+        if token_tuple[0] is not None:
+
+            # There is already a block
+            blocks.append(id_to_cluster[token_tuple[0]])
+
+        else:
+
+            # Case of multiple surnames
+            tokens = token_tuple[1]
+            surnames, first_names = tokens
+            last_metaphone_score = 0
+
+            # Check if this combination of surnames was already included
```

Comment:
For further reference, we should mention here that this makes the blocking ordering dependent (which is OK, but we should not forget that)